src/algorithms/davidchack_lai.jl

```diff
@@ -0,0 +1,201 @@
+        end
+
+        if !(1 <= m <= n)
+            throw(ArgumentError("`m` must be in [1, `n`=$(n)]"))
+        end
+        return new(n, m, β, maxiters, disttol, abstol)
+    end
+end
+
+function periodic_orbits(ds::DeterministicIteratedMap, alg::DavidchackLai, igs::Vector{InitialGuess})
+    if isinplace(ds)
+        throw(ArgumentError("Algorithms `DavidchackLai` currently supports only out of place systems."))
+    end
+
+    type = typeof(current_state(ds))
+    fps = storage(type, alg.n)
+    igs = [ig.u0 for ig in igs]
+
+    isnothing(alg.β) ? β = n-> 10*1.2^(n) : β = alg.β
+    indss, signss = lambdaperms(dimension(ds))
+    C_matrices = [cmatrix(inds,signs) for inds in indss, signs in signss]
+
+    initial_detection!(fps, ds, alg, igs, β, C_matrices)
+    main_detection!(fps, ds, alg, β, C_matrices)
+
+    return uniquepos(output(ds, fps, type, alg.n); atol=alg.abstol)
+end
+
+function initial_detection!(fps, ds, alg, igs, β, C_matrices)
+    for i in 1:alg.m
+        detect_orbits!(fps[i], ds, alg, i, igs, β(i), C_matrices)
+    end
+end
+
+function main_detection!(fps, ds, alg, β, C_matrices)
+    for period = 2:alg.n
+        previousfps = fps[period-1]
+        currentfps = fps[period]
+        nextfps = fps[period+1]
+        for (container, seed, period) in [
+            (currentfps, previousfps, period), 
+            (nextfps, currentfps, period+1), 
+            (currentfps, nextfps, period)
+            ]
+            detect_orbits!(container, ds, alg, period, seed, β(period), C_matrices)
+        end
+    end
+end
+
+function _detect_orbits!(fps, ds, alg, T, igs, C, β)
+    for x in igs
+        x = x
+        for _ in 1:(isnothing(alg.maxiters) ? max(100, 4*β) : alg.maxiters)
+            xn = DL_rule(x, β, C, ds, T)
+            if norm(g(ds, xn, T)) < alg.disttol
+                if !iscontained(xn, fps, alg.abstol)
+                    push!(fps, xn)
+                    break
+                end
+            end
+            x = xn
+        end
+    end
+end
+
+function detect_orbits!(fps, ds, alg, T, igs, β, C_matrices)
+    for C in C_matrices
+        _detect_orbits!(fps, ds, alg, T, igs, C, β)
+    end
+end
+
+function DL_rule(x, β, C, ds, n)
+    Jx = DynamicalSystemsBase.ForwardDiff.jacobian(x0 -> g(ds, x0, n), x)
+    gx = g(ds, x, n)
+    xn = x + inv(β*norm(gx)*C' - Jx) * gx
+    return xn
+end
+
+function g(ds, state, n)
+    p = current_parameters(ds)
+    newst = state
+    # not using step!(ds, n) to allow automatic jacobian
+    for _ = 1:n
+        newst = ds.f(newst, p, 1.0)
+    end
+    return newst - state
+end
+
+function storage(type, n)
+    vectype = Vector{type}
+    storage = Vector{vectype}(undef, n+1)
+    for i in 1:n+1
+        storage[i] = vectype[]
+    end
+    return storage
+end
+
+function iscontained(x, arr, thresh)
+    for y in arr
+        if norm(x - y) < thresh
+            return true
+        end
+    end
+    return false
+end
+
+function output(ds, fps, type, T)
```

Comment:
This function has an incredibly generic name. It should be named specifically to what it does. Literally every function can be named "output" as it does return an "output".